app/src/ui/dialog/dialog.tsx

```diff
@@ -221,9 +220,135 @@ export class Dialog extends React.Component<IDialogProps, IDialogState> {
+   *   a harmless action. [...] when a dialog may result in a destructive
+   *   action, Cancel can be set as the default button."
+   *
+   * The same guidelines also has this to say about focus:
+   *
+   *   "Set the initial focus to the first location that accepts user input.
+   *    Doing so lets the user begin entering data immediately, without needing
+   *    to click a specific item like a text field or list."
+   *
+   * In attempting to follow the guidelines outlined above we follow a priority
+   * order in determining the first suitable child.
+   *
+   *  1. The element with the lowest positive tabIndex
+   *     This might sound counterintuitive but imagine the following pseudo
+   *     dialog this would be button D as button D would be the first button
+   *     to get focused when hitting Tab.
+   *
+   *     <dialog>
+   *      <button>A</button>
+   *      <button tabIndex=3>B</button>
+   *      <button tabIndex=2>C</button>
+   *      <button tabIndex=1>D</button>
+   *     </dialog>
+   *
+   *  2. The first element which is either implicitly keyboard focusable (like a
+   *     text input field) or explicitly focusable through tabIndex=0 (like a TabBar
+   *     tab)
+   *
+   *  3. The first submit button. We use this as a proxy for what macOS HIG calls
+   *     "default button". It's not the same thing but for our purposes it's close
+   *     enough.
+   *
+   *  4. Any remaining button
+   *
+   */
+  private focusFirstSuitableChild() {
+    const dialog = this.dialogElement
+
+    if (dialog === null) {
+      return
+    }
+
+    const selector = [
+      'input:not([type=hidden]):not(:disabled):not([tabindex="-1"])',
+      'textarea:not(:disabled):not([tabindex="-1"])',
+      'button:not(:disabled):not([tabindex="-1"])',
+      '[tabindex]:not(:disabled):not([tabindex="-1"])',
+    ].join(', ')
+
+    // The element which has the lowest explicit tab index (i.e. greater than 0)
+    let firstExplicit: { 0: number; 1: HTMLElement | null } = [Infinity, null]
+
+    // First submit button
+    let firstSubmitButton: HTMLElement | null = null
+
+    // The first button-like element (input, submit, reset etc)
+    let firstButton: HTMLElement | null = null
+
+    // The first element which is either implicitly keyboard focusable (like a
+    // text input field) or explicitly focusable through tabIndex=0 (like an
+    // anchor tag masquerading as a button)
+    let firstTabbable: HTMLElement | null = null
+
+    const excludedInputTypes = [
+      ':not([type=button])',
+      ':not([type=submit])',
+      ':not([type=reset])',
+      ':not([type=hidden])',
+      ':not([type=checkbox])',
+      ':not([type=radio])',
+    ]
+
+    const inputSelector = `input${excludedInputTypes.join('')}, textarea`
+    const buttonSelector =
+      'input[type=button], input[type=submit] input[type=reset], button'
+    const submitSelector = 'input[type=submit], button[type=submit]'
```

Comment:
these could be defined at the file-level as well, i think? (this is a nit-pick and is non-blocking)